src/buffer.c

```diff
@@ -109,44 +112,115 @@ free_buffer(struct Buffer *buf) {
 
 ssize_t
 buffer_recv(struct Buffer *buffer, int sockfd, int flags, struct ev_loop *loop) {
+    struct msghdr msg = { 0 };
+
+    return buffer_recvmsg(buffer, sockfd, &msg, flags, loop);
+}
+
+ssize_t
+buffer_recvmsg(struct Buffer *buffer, int sockfd, struct msghdr *msg,
+        int flags, struct ev_loop *loop) {
     /* coalesce when reading into an empty buffer */
     if (buffer->len == 0)
         buffer->head = 0;
 
     struct iovec iov[2];
-    struct msghdr msg = {
-        .msg_iov = iov,
-        .msg_iovlen = setup_write_iov(buffer, iov, 0)
-    };
+    ssize_t bytes;
+
+    if (buffer->type == SOCK_DGRAM) {
+        uint16_t *dgram_len = reserve_dgram_length(buffer);
```

Comment:
We need to check this result in case there isn't free space in the dgram buffer. Sorry I missed it in my prototype.